---
title: "Power Ranges : GenomicRanges workbook"
output: html_notebook
---


```{r setup}
library(R.utils)
library(SummarizedExperiment)
library(Rsamtools)
library(qs)
library(purrr)
library(data.table)
library(stringi)
```


```{r fetch example files}
options(timeout = 300) # default is 60 -- genome download times out
dlurls=list('hg38.fa.bgz' = 'EfWx9pg_ijlOqGUZZBI1FXQBpOOBuvhE-KvoZLWXQdOYSw?e=tVY4tL',
            'hg38.fa.bgz.fai' = 'EUQr4eDAJVtGv-UY2p9wOnsBcU3_Cmdk-ysIVgXyqrOxTw?e=FKEGJY',
            'hg38.fa.bgz.gzi' = 'EVWjuVzLc2hFg8ZPLO6RzscBBeqllusCA7BssBer1qvB9g?e=Zex951',
            'rse_jx.qs' = 'ERm6QKac4F5Hkb34dleefXwBBZqbN9ixmaAX9J6LBdsB8w?e=asYSWr',
            'gencode43.nri.main.gtf.gz' = 'EcrHEjtDsI5IrxLQ74HharIBroPzsEOWqMzW_2CMMgk9yQ?e=XtQ82j'
)
jhl='https://livejohnshopkins-my.sharepoint.com/:u:/g/personal/gpertea1_jh_edu/'
dl='&download=1'; datadir='data/'
if (!dir.exists(datadir)) dir.create(datadir)
for (file in names(dlurls)) {
  dlto=paste0(datadir,file)
  if (!file.exists(dlto)) {
   message("downloading: ",dlto)
   download.file(paste0(jhl,dlurls[[file]],dl), destfile=dlto, mode = 'wb')
  } else {
    message(" file ", dlto, " already exists." )
  }
}

## last file has the Gencode annotation
anngz=paste0(datadir, names(dlurls)[length(dlurls)])
```

```{r load/import}
annqs=sub('\\.gz$','.qs',anngz)
if (file.exists(annqs)) {
  anndata <- qread(annqs)          ##  ~ 4 seconds
  #load(sub('qs$','Rdata', annqs)) ##  ~ 9 seconds
} else {
  anndata <- rtracklayer::import(anngz) ## ~20 seconds
  qsave(anndata, file=annqs)            ## ~ 5 seconds
  # save(anndata, file=sub('qs$','Rdata', annqs)) ## ~ 12 seconds
}
## file.edit('~/.Rprofile') ## defined oi, os, hh, etc.
## oi(anndata)
#anndata
table(anndata$type)
```

```{r get exonic length for transcripts, intron details}
anntxgr <- anndata[anndata$type=='transcript'] ## transcript GRanges only
names(anntxgr) <- anntxgr$transcript_id ## set names so we can select GRanges by transcript_id
annexgr <- anndata[anndata$type=='exon'] ## exon GRanges only
## split: create a GRangesList with exons grouped by transcript
txgrl <- split(annexgr, annexgr$transcript_id)
stopifnot(identical(length(txgrl), length(anntxgr))) # sanity check

## GRangesList::width returns an IntegerList with names as transcript_id
## summing up exon length per all transcripts:
txlens <- sum(width(txgrl)) # named list of exonic lengths for each transcript!
head(txlens)
## add new metadata column with the exonic length:
anntxgr$Length <- txlens[anntxgr$transcript_id]
anntxgr$numExons <- lengths(txgrl) ## or elementNROWS()
### so now anntxgr has both Length and numExons

## show ordered by decreasing number of exons - titin gene!
anntxgr[order(anntxgr$numExons, decreasing = T)]

## -- get intron GRanges for each transcript 
## only multiexon transcripts have exons
metxgrl <- txgrl[lengths(txgrl)>1] # 190689 vs 216121 total
metxgr <- anntxgr[names(metxgrl)]
stopifnot(identical(names(metxgr), names(metxgrl)))
## get introns using psetdiff: subtract exon ranges from transcript ranges!
introns <- psetdiff(metxgr, metxgrl) ## introns from every multi-exon transcript
introns <- unlist(introns)
introns$transcript_id <- names(introns)
annJxData <- NULL
annjxqs=sub('\\.gz$','.annjx.qs',anngz)
## we're going to use this prepared annotation file for novel junction annotation
if (file.exists(annjxqs)) { 
  annJxData <- qread(annjxqs)
} else {
  ## - data.table has better aggregation features
  ## as we want to keep lists of genes and transcripts for each junction
  annJxData <- as.data.table(introns)
  jgenes <- mcols(metxgr[annJxData$transcript_id])[, c('gene_id', 'gene_name')]
  ## add gene IDs, collapse identical junctions but keep lists of transcripts, genes sharing them
  annJxData <- annJxData[, `:=`(gene_id=jgenes$gene_id, gene_name=jgenes$gene_name)][,
               .( txlist=list(sort(unique(c(transcript_id)))), glist=list(sort(unique(c(gene_id)))),
              gsyms=list(sort(unique(c(gene_name)))) ), by=.(seqnames, start, end, strand)][,
                                `:=`(gcount=lengths(glist), txcount=lengths(txlist)) ]
  annJxData <- makeGRangesFromDataFrame(annJxData, keep.extra.columns = T)
  names(annJxData) <-  paste(seqnames(annJxData), ranges(annJxData), strand(annJxData), sep=':')
  qsave(annJxData, file=annjxqs)
}
annJxData

```


```{r get exonic length of genes}
## -- adding Length to genes: we need to merge exons first
##  split:  create a GRangesList with exons grouped by gene
## reduce:  merge overlapping exons within each gene group
mrgrl <- GenomicRanges::reduce(split(annexgr, annexgr$gene_id))
genelens <- sum(width(mrgrl)) ## => named list of exonic lengths for each gene
## maintain a gene GRanges
anngene <- anndata[anndata$type=='gene']
anngene$Length <- genelens[anngene$gene_id]

```

```{r load junctions}
rsejx <- qread(paste0('data/',names(dlurls)[4]))
jxgr <- rowRanges(rsejx)
## let's pretend they are not annotated yet
mcols(jxgr) <- NULL ## remove metadata
names(jxgr) <- NULL ## remove old names
## assign unique names from genomic location
names(jxgr) <- stri_join(seqnames(jxgr),ranges(jxgr), strand(jxgr), sep=':')
jxgr
```


```{r annotate junctions}
## annotate junctions vs loaded reference annotation annJxData
## check these junction against the known intron database we built earlier
inGencode <- countOverlaps(jxgr, annJxData, type = "equal") > 0
jxgr$inGencode <- inGencode
message(length(inGencode[inGencode]), " junctions found matching annotation.")
#if (length(inGencode)<100000) {
#  message("WARNING: low number of Gencode-matching junctions, check for coordinate/strand issues")
#}

jxgr$inGencodeStart <- countOverlaps(jxgr, annJxData, type = "start") > 0
jxgr$inGencodeEnd <- countOverlaps(jxgr, annJxData, type = "end") > 0

oo <- findOverlaps(jxgr, annJxData, type = "equal")

# list of Gencode IDs fully matching junction
pwalk(list(c('geneids', 'symbols', 'txlist'),
           c('glist','gsyms','txlist'),
           c('gencodeGeneID','Symbol','gencodeTx')),
      function(a, b, c) {
        dt <- data.table(l=vector('list', length(jxgr)), s=as.character(NA))
        ## flatten the list columns into comma-delimited strings
        dt[queryHits(oo), l:=mcols(annJxData[subjectHits(oo)])[, b] ]
        dt[lengths(l)>1,  s:=sapply(l, paste, collapse=',')]
        dt[lengths(l)==1, s:=as.character(l)]
        mcols(jxgr)[, c] <<- dt[,2]
        if (b=='txlist') {
          jxgr$numTx <<- lengths(dt$l)
        }
      })
## junction code
jxgr$Class <- ifelse(jxgr$inGencode, "InGen",
           ifelse(jxgr$inGencodeStart & jxgr$inGencodeEnd, "ExonSkip",
            ifelse(jxgr$inGencodeStart | jxgr$inGencodeEnd, "AltStartEnd", "Novel")
                     )
)
## -- more annotations performed in a similar way, also detecting "fusions" etc.
## ... rest of the code not included here ...
jxgr
```

```{r annotate novel transcript}
## let's define an "unkown" transcript with 3 exons
unktx <- GRanges(seqnames=Rle('chr11', 3), 
              ranges=IRanges(start=c(130910018,130906625,130905953), 
                             end=c(130910137,130906721,130906133)), strand='-')
unktx$transcript_id <- 'unkTx'
names(unktx) <- unktx$transcript_id
## report any overlaps between exons
hh <- findOverlaps(unktx, anntxgr)
hh ## show overlaps for every exon

## create a GRangesList for this transcript
unktxl <- split(unktx, unktx$transcript_id) 
## findOverlaps between GRangesList objects - parent level hit reporting!
hl <- findOverlaps(unktxl, txgrl)
hl
## we can add metadata columns at parent level in GRangesList !
## this column collects all genes of transcripts overlapping our unknown transcripts
mcols(unktxl)$genes <- list(unique(anntxgr[names(txgrl[subjectHits(hl)])]$gene_name))
unktxl;mcols(unktxl)
## we can also add the "genes" annotation at exon level:
# mcols(unktx)$genes <- mcols(unktxl)$genes
```

```{r write intergenic sequences}
## open the genome file for reading (Rsamtools)
## this also opens and loads the index, which has the chromosome lengths
gfa <- 'data/hg38.fa.bgz'

fa <- open(FaFile(gfa)) #seqinfo(fa) has the chromosome sequences info

## make a GRange for the whole chromosome 7 (strand is not important)
cgr <- GRanges(seqnames=seqnames(seqinfo(fa)), ranges = IRanges(start=1, end=seqlengths(seqinfo(fa))))
names(cgr) <- seqnames(cgr)

usgr <- subset(anndata, type=='gene')
strand(usgr)='*' # make the gene ranges unstranded
## get the merged ranges of overlapping genes on each chromosome
mrgrl <- GenomicRanges::reduce(split(usgr, seqnames(usgr)))
cgr <- cgr[names(mrgrl)] #make sure they have the same order

## get intergenic GRanges by subtracting gene GRanges from chromosome GRanges
intergr <- psetdiff(cgr, mrgrl) # getting intergenic GRanges

## write the intergenic region sequences for chr7 only:
igseq <- getSeq(fa, intergr[['chr7']])
names(igseq) <- paste0(names(igseq), '_ig', sprintf('%04d', 1:length(igseq)))
writeXStringSet(igseq, filepath = 'chr7_intergenics.fa', append=F)

```


```{r write transcript sequences}
## --- write transcript sequences ----------------
## get the exon GRanges for transcripts in the last quarter of chromosome 9 
tx9 <- anndata[anndata$type=='transcript' & seqnames(anndata)=='chr9']

# minimum coordinate for transcript in the last 1/4 of chromosome 9
cstart <- 3*(seqlengths(seqinfo(fa)['chr9'])/4)
tx9 <- tx9[start(tx9)>=cstart] # 3178

## get just the exons for these selected transcripts
tx9exons <- anndata[anndata$transcript_id %in% tx9$transcript_id & 
                    anndata$type=='exon']

## get a GRangesList with exons grouped by transcript_id
txgrl <- split(tx9exons, tx9exons$transcript_id) 

txchunk <- DNAStringSet() # chunk of transcript sequences to be written to FASTA
numtx <- 0
fappend <- F
fasta <- 'transcripts_chr9_subset.fa'

for (tid in names(txgrl)) {
    txExonRanges <- txgrl[[tid]] # get the exons for the transcript
    txseq <- DNAStringSet(unlist(getSeq(fa, txExonRanges)))
    names(txseq) <- tid
    numtx <- numtx+1
    txchunk <- c(txchunk, txseq)
    if (numtx %% 1100 == 0) {
       writeXStringSet(txchunk, fasta, append=fappend)
       cat('..', numtx)
       fappend <- T
       txchunk <- DNAStringSet() 
    }
}
if (length(txchunk)>0) {
   writeXStringSet(txchunk, fasta, append=fappend)
   cat('..', numtx)
}
cat(" transcript sequences written.\n")
close(fa)
```

```{r}

```

